File: src/main/java/com/company/commitet_jm/sheduledJob/GitCloneTask.java
package com.company.commitet_jm.sheduledJob;

import com.company.commitet_jm.service.git.GitService;
import io.jmix.core.DataManager;
import io.jmix.core.FileStorageLocator;
import io.jmix.flowui.backgroundtask.BackgroundTask;
import io.jmix.flowui.backgroundtask.TaskLifeCycle;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

public class GitCloneTask extends BackgroundTask<Integer, Void> {
    private static final org.slf4j.Logger log = LoggerFactory.getLogger(GitCloneTask.class);

    private final DataManager dataManager;
    private final FileStorageLocator fileStorageLocator;
    private final GitService gitService;

    private String urlRepo = "";
    private String localPath = "";
    private String defaultBranch = "";

    public GitCloneTask(
            DataManager dataManager,
            FileStorageLocator fileStorageLocator,
            GitService gitService
    ) {
        super(10, TimeUnit.MINUTES);
        this.dataManager = dataManager;
        this.fileStorageLocator = fileStorageLocator;
        this.gitService = gitService;
    }

    @Override
    public Void run(TaskLifeCycle<Integer> taskLifeCycle) throws Exception {
        GitWorker.Pair<Boolean, String> result = gitService.cloneRepo(urlRepo + ".git", localPath, defaultBranch);

        if (!result.getFirst()) {
            log.error("Ошибка клонирования: " + result.getSecond());
        } else {
            log.info("репозиторий склонирован");
        }

        return null;
    }

    // Getters and setters
    public String getUrlRepo() {
        return urlRepo;
    }

    public void setUrlRepo(String urlRepo) {
        this.urlRepo = urlRepo;
    }

    public String getLocalPath() {
        return localPath;
    }

    public void setLocalPath(String localPath) {
        this.localPath = localPath;
    }

    public String getDefaultBranch() {
        return defaultBranch;
    }

    public void setDefaultBranch(String defaultBranch) {
        this.defaultBranch = defaultBranch;
    }
}

File: bin/main/com/company/commitet_jm/sheduledJob/GitCloneTask.kt
package com.company.commitet_jm.sheduledJob

import com.company.commitet_jm.entity.Project
import com.company.commitet_jm.service.project.ProjectService
import io.jmix.flowui.backgroundtask.BackgroundTask
import io.jmix.flowui.backgroundtask.TaskLifeCycle
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit

/**
 * Фоновая задача для клонирования репозитория проекта
 */
open class GitCloneTask(
    private val projectService: ProjectService,
    private val project: Project
) : BackgroundTask<Int, String>(10, TimeUnit.MINUTES) {

    companion object {
        private val log = LoggerFactory.getLogger(GitCloneTask::class.java)
    }

    /**
     * Результат клонирования
     */
    var resultPath: String? = null
    var errorMessage: String? = null

    @Throws(Exception::class)
    override fun run(taskLifeCycle: TaskLifeCycle<Int>): String {
        log.info("Начало клонирования репозитория для проекта: ${project.name}")

        val result = projectService.cloneProjectRepo(project)

        if (taskLifeCycle.isCancelled) {
            log.info("Задача клонирования отменена")
            return "Отменено"
        }

        if (result.first) {
            resultPath = result.second
            log.info("Репозиторий успешно склонирован: $resultPath")
            return resultPath!!
        } else {
            errorMessage = result.second
            log.error("Ошибка клонирования: $errorMessage")
            throw RuntimeException(errorMessage)
        }
    }
}
